mpc-core/src/protocols/rep3/yao.rs

```diff
@@ -0,0 +1,481 @@
+        let mut res = Vec::with_capacity(n_bits);
+        let mut bits = 0;
+        for mut el in input.to_u64_digits() {
+            for _ in 0..64 {
+                res.push((el & 1) as u16);
+                el >>= 1;
+                bits += 1;
+                if bits == n_bits {
+                    break;
+                }
+            }
+        }
+        res.resize(n_bits, 0);
+        res
+    }
+
+    /// Encode a wire, producing the zero wire as well as the encoded value.
+    pub fn encode_wire<R: Rng + CryptoRng>(
+        rng: &mut R,
+        delta: &WireMod2,
+        val: u16,
+    ) -> (WireMod2, WireMod2) {
+        let zero = WireMod2::rand(rng, 2);
+        let enc = zero.plus(&delta.cmul(val));
+        (zero, enc)
+    }
+
+    /// This puts the X_0 values into garbler_wires and X_c values into evaluator_wires
+    fn encode_bits<R: Rng + CryptoRng>(
+        bits: Vec<u16>,
+        rng: &mut R,
+        delta: WireMod2,
+    ) -> GCInputs<WireMod2> {
+        let mut garbler_wires = Vec::with_capacity(bits.len());
+        let mut evaluator_wires = Vec::with_capacity(bits.len());
+        for bit in bits {
+            let (mine, theirs) = Self::encode_wire(rng, &delta, bit);
+            garbler_wires.push(mine);
+            evaluator_wires.push(theirs);
+        }
+        GCInputs {
+            garbler_wires: BinaryBundle::new(garbler_wires),
+            evaluator_wires: BinaryBundle::new(evaluator_wires),
+            delta,
+        }
+    }
+
+    /// This puts the X_0 values into garbler_wires and X_c values into evaluator_wires
+    pub fn encode_bigint<R: Rng + CryptoRng>(
+        bigint: &BigUint,
+        n_bits: usize,
+        rng: &mut R,
+        delta: WireMod2,
+    ) -> GCInputs<WireMod2> {
+        let bits = Self::biguint_to_bits_as_u16(bigint, n_bits);
+        Self::encode_bits(bits, rng, delta)
+    }
+
+    /// This puts the X_0 values into garbler_wires and X_c values into evaluator_wires
+    pub fn encode_field<F: PrimeField, R: Rng + CryptoRng>(
+        field: F,
+        rng: &mut R,
+        delta: WireMod2,
+    ) -> GCInputs<WireMod2> {
+        let bits = Self::field_to_bits_as_u16(field);
+        Self::encode_bits(bits, rng, delta)
+    }
+}
+
+/// Transforms an arithmetically shared input x = (x_1, x_2, x_3) into three yao shares x_1^Y, x_2^Y, x_3^Y. The used delta is an input to the function to allow for the same delta to be used for multiple conversions.
+pub fn joint_input_arithmetic<F: PrimeField, N: Rep3Network, R: Rng + CryptoRng>(
+    x: Rep3PrimeFieldShare<F>,
+    delta: Option<WireMod2>,
+    io_context: &mut IoContext<N>,
+    rng: &mut R,
+) -> IoResult<[BinaryBundle<WireMod2>; 3]> {
+    let id = io_context.id;
+    let n_bits = F::MODULUS_BIT_SIZE as usize;
+
+    // x1 is known by both garblers, we can do a shortcut
+    let mut x1 = (0..n_bits)
+        .map(|_| WireMod2::from_block(io_context.rngs.generate_shared::<Block>(id), 2))
```

Comment:
Is generate_shared ok here, since the remaining party also has this value?

Reply:
so the idea is that x1 is what the evaluator gets, i.e., zero_wire + x * delta. So the evaluator does not know the share. x1 is known by both garblers in advance, so they can produce the zero_wire and the one_wire from it.

See also https://eprint.iacr.org/2018/403.pdf p18, last paragraph of "Joint Yao Input"

Reply:
Ok I see so this is just to save sending the generated x1 to the evaluator. Maybe put link in comment above.

Reply:
done